src/dataflow/src/source/kafka.rs

```diff
@@ -656,172 +810,220 @@ where
+                    let offset = MzOffset::from(message.offset);
 
-                partition_metrics.offset_received.set(offset.offset);
+                    partition_metrics.offset_received.set(offset.offset);
 
-                // Determine the timestamp to which we need to assign this message
-                let ts = find_matching_timestamp(&id, partition, offset, &timestamp_histories);
-                match ts {
-                    None => {
-                        // We have not yet decided on a timestamp for this message,
-                        // we need to buffer the message
-                        dp_info.buffer_message(message);
+                    // Determine the timestamp to which we need to assign this message
+                    let ts = find_matching_timestamp(
+                        &mut cp_info,
+                        &id,
+                        partition,
+                        offset,
+                        &timestamp_histories,
+                    );
+                    match ts {
+                        None => {
+                            // We have not yet decided on a timestamp for this message,
+                            // we need to buffer the message
+                            dp_info.buffer_message(message);
+                            activator.activate();
+                            return SourceStatus::Alive;
+                        }
+                        Some(ts) => {
+                            // Note: empty and null payload/keys are currently
+                            // treated as the same thing.
+                            let key = message.key.unwrap_or_default();
+                            let out = message.payload.unwrap_or_default();
+                            // Entry for partition_metadata is guaranteed to exist as messages
+                            // are only processed after we have updated the partition_metadata for a
+                            // partition and created a partition queue for it.
+                            cp_info
+                                .partition_metadata
+                                .get_mut(&partition)
+                                .unwrap()
+                                .offset = offset;
+                            bytes_read += key.len() as i64;
+                            bytes_read += out.len() as i64;
+                            let ts_cap = cap.delayed(&ts);
+                            output.session(&ts_cap).give(SourceOutput::new(
+                                key,
+                                out,
+                                Some(Into::<KafkaOffset>::into(offset).offset),
+                            ));
+
+                            partition_metrics.offset_ingested.set(offset.offset);
+                            partition_metrics.messages_ingested.inc();
+
+                            downgrade_capability(
```

Comment:
For RT sources at least (I haven't looked through the BYO code path yet), this calls `cp_info.generate_next_timestamp()`, but that has already been called by `find_matching_timestamp`. Is it right that we call that twice for the same message?

Reply:
The "goal" is to dowgrade the capability to the highest possible value so as to keep the stream as "fresh" as possible (when joining against other streams), hence why I downgrade to a new timestamp.